src/peergos/shared/user/fs/Capability.java

```diff
@@ -14,49 +14,68 @@
 import java.util.stream.Stream;
 
 public class Capability implements Cborable {
-    public final Location location;
+    // writer is only present when it is not implicit (an entry point, or a child link to a different writing key)
+    public final Optional<PublicKeyHash> writer;
+    private final byte[] mapKey;
     public final SymmetricKey baseKey;
-    public final Optional<SecretSigningKey> writer;
+    public final Optional<SecretSigningKey> signer;
 
     @JsConstructor
-    public Capability(Location location, Optional<SecretSigningKey> writer, SymmetricKey baseKey) {
-        this.location = location;
-        this.baseKey = baseKey;
+    public Capability(Optional<PublicKeyHash> writer, byte[] mapKey, SymmetricKey baseKey, Optional<SecretSigningKey> signer) {
         this.writer = writer;
+        if (mapKey.length != Location.MAP_KEY_LENGTH)
+            throw new IllegalStateException("Invalid map key length: " + mapKey.length);
+        this.mapKey = mapKey;
+        this.baseKey = baseKey;
+        this.signer = signer;
+    }
+
+    public Capability(byte[] mapKey, SymmetricKey baseKey) {
+        this(Optional.empty(), mapKey, baseKey, Optional.empty());
+    }
+
+    public Capability(PublicKeyHash writer, byte[] mapKey, SymmetricKey baseKey, Optional<SecretSigningKey> signer) {
+        this(Optional.of(writer), mapKey, baseKey, signer);
+    }
+
+    public Capability(PublicKeyHash signer, byte[] mapKey, SymmetricKey baseKey) {
+        this(Optional.of(signer), mapKey, baseKey, Optional.empty());
     }
 
-    public Capability(PublicKeyHash owner, PublicKeyHash writer, byte[] mapKey, SymmetricKey baseKey) {
-        this(new Location(owner, writer, mapKey), Optional.empty(), baseKey);
+    public Capability(SigningPrivateKeyAndPublicHash signer, byte[] mapKey, SymmetricKey baseKey) {
+        this(Optional.of(signer.publicKeyHash), mapKey, baseKey, Optional.of(signer.secret));
     }
 
-    public Capability(PublicKeyHash owner, SigningPrivateKeyAndPublicHash writer, byte[] mapKey, SymmetricKey baseKey) {
-        this(new Location(owner, writer.publicKeyHash, mapKey), Optional.of(writer.secret), baseKey);
+    @JsMethod
+    public byte[] getMapKey() {
+        return Arrays.copyOf(mapKey, mapKey.length);
     }
 
-    public Location getLocation() {
-        return location;
+    public Location getLocation(PublicKeyHash owner, PublicKeyHash writer) {
+        return new Location(owner, this.writer.orElse(writer), mapKey);
```

Comment:
> this.writer.orElse(writer)

I've been thinking about this for a few minutes. This method seems like a bug waiting to happen. What if instead we it split into two explicit methods: one for the entry-point case and one for the other case (where this.writer is empty)?

Reply:
What is the bug you are thinking of? This is only called from one place currently: in a directory when we are retrieving all the children. In that case we need to use the directory's writer key, unless the value here is present (the child is under a different writing key).